src/pass.h

```diff
@@ -123,12 +123,22 @@ struct PassRunner {
 
   ~PassRunner();
 
+  virtual bool isNested() { return false; }
+
 private:
   void doAdd(Pass* pass);
 
   void runPassOnFunction(Pass* pass, Function* func);
 };
 
+// A pass runner that is safe to run inside a pass
+struct NestedPassRunner : public PassRunner {
```

Comment:
Oh nice, this was pretty much what I was thinking of. Wound up being less cross-cutting than I'd thought.